src/core/Akka.Streams/Implementation/IO/FileSubscriber.cs

```diff
@@ -105,113 +117,158 @@ protected override void PreStart()
         {
             try
             {
-                _chan = _f.Open(_fileMode, FileAccess.Write, FileShare.ReadWrite);
+                _chan = new FileStream(_f.ToString(), _fileMode, FileAccess.Write, FileShare.ReadWrite, _fileBufferSize, true);
                 if (_startPosition > 0)
                     _chan.Position = _startPosition;
-                base.PreStart();
             }
             catch (Exception ex)
             {
-                CloseAndComplete(new Try<IOResult>(ex));
+                _completionPromise.SetException(ex);
                 Cancel();
             }
         }
 
         /// <summary>
         /// TBD
         /// </summary>
-        /// <param name="message">TBD</param>
-        /// <returns>TBD</returns>
         protected override bool Receive(object message)
         {
             switch (message)
             {
-                case OnNext next:
-                    try
+                case OnNext next when _fileRequests == 0:
                     {
-                        var byteString = (ByteString)next.Element;
-                        var bytes = byteString.ToArray();
-                        _chan.Write(bytes, 0, bytes.Length);
-                        _bytesWritten += bytes.Length;
-                        if (_autoFlush)
-                            _chan.Flush(true);
+                        var bytes = (ByteString)next.Element;
+                        if (_bufferedByteString.Count > 0)
+                        {
+                            bytes = _bufferedByteString.Concat(bytes);
+                            _bufferedByteString = ByteString.Empty;
+                            _requestStrategy.FileBufferCount = 0;
+                        }
+                        _streamTask = WriteAsync(Self, bytes, _autoFlush);
+                        _fileRequests++;
```

Comment:
Is it possible that `_fileRequest` be greater than 1? If not, is it better to change this into a `bool` instead, or just check and see if `_streamTask` is null, since `_streamTask` has to be assigned if a write operation is currently in progress?

Reply:
`_streamTask` is not used (maybe for debug), because `_streamTask.IsCompleted` is not writable by the actor itself,
_fileRequests can only be 0 or 1 for this actor, but this pattern can be used for other things
its very good for fan-out / fad-in to async operations

Reply:
But does fan-out make any sense for a file sink? This should be a sink that does a write to a single file, and not multiple files.

Reply:
Should i convert `int _fileRequests` to `boolean _fileRequested` ?

Reply:
I think so, making it a bool would make the intention more clear. Or we can use the _streamTask as a flag to check if a write is progress or not.

Reply:
`_streamTask.IsCompleted` can not be used, because it would pre-indicate inflight completion messages.
I will change the state field to a bool.

Reply:
Well, you can always check to see if `_streamTask` is null or not, set it to null whenever a write operation is completed.